dataset/jeopardy_dataset.py

```diff
@@ -0,0 +1,275 @@
+import os.path
+import json
+from PIL import Image
+from torch.utils.data import Dataset
+import torch
+import os
+import string
+from tqdm import tqdm
+import nltk
+
+# START_TOKEN = "<s>"
+# END_TOKEN = "</s>"
+PAD_TOKEN = "<pad>"
+UNK_TOKEN = "<unk>"
+
+class JeopardyDataset(Dataset):
+    def __init__(self, 
+        questions_file, 
+        answers_file, 
+        images_dir, 
+        transform, 
+        word2idx=None, 
+        train=True, 
+        q_len=8, 
+        ans_len=2, 
+        test_split=0.2, 
+        dumb_transfer=False, 
+        most_common_answers=None, 
+        num_answers=0, 
+        frequency_threshold=8, 
+        multiple_images=False):
+        """
+        Args:
+            questions_file (string): Path to the json file with questions.
+            answers_file (string): Path to the json file with annotations.
+            images_dir (string): Directory with all the images.
+            transform (callable, optional): Optional transform to be applied
+                on a sample.
+            word2idx: word2idx[word] = index, constructed on the train set only, must be passed in for the test set.  
+            frequency_threshold: how many times does a word have to appear as an answer for it to be considered. (default: 8) 
+            As used in Teney et al. https://arxiv.org/pdf/1708.02711.pdf
+                
+        Example call:
+            test_ds = JeopardyDataset("v2_OpenEnded_mscoco_train2014_questions.json", "v2_mscoco_train2014_annotations.json", "images")
+
+        """
+        self.dumb_transfer = dumb_transfer
+        self.multiple_images = multiple_images
+        # initializing the lengths of our questions and answers
+        self.q_len = q_len
+        self.ans_len = ans_len
+        self.num_answers = num_answers
+        # return test or train set?
+        self.train = train
+        self.glove_dict = self._get_glove_indices()
+
+        q_f = open(questions_file, 'r')
+        a_f = open(answers_file, 'r')
+        
+        # we want only 80% of the question/answer text to build the vocabulary
+        self.answers = json.load(a_f)["annotations"]
+        self.questions = json.load(q_f)["questions"]
+
+        split_index = int((1-test_split) * len(self.questions))
```

Comment:
make sure the questions aren't sorted in any special way where if we take the last `test_split` as test , they are all of a certain class or something

Reply:
ah great point

Reply:
this is fixed now btw, I randomize the questions when saving the file

Reply:
but why are we randomizing questions? In general, I prefer to cache as few things as possible as it induces bugs very often.